### DISCUSSION TITLE
Adding experiment to the models created from scratch

### DISCUSSION THREAD

**Discussion by manjunathnilugal** [pybamm-team/PyBaMM#3341] [Q&A] [answered]
Hi everyone, 

I want to know how to **_add experiments in the **_models that are created from scratch_**_**.
I have a model working for constant current and voltage.
Is there any method in Pybamm to implement experiment to implement CCCV in cycles  that are build from scratch?

Regards,
Manjunath

**Comment by brosaplanella** [comment]
PyBaMM uses the [external circuit submodels](https://docs.pybamm.org/en/latest/source/api/models/submodels/external_circuit/index.html) for experiments, so it is quite tricky to get this working with models that are not integrated with the PyBaMM structure. It might be easier to
a) Write the additional constraint to do CCCV yourself
b) Try to integrate your model in the PyBaMM structure

What's your model? Depending on this, b) might be quite easy.

**Reply by manjunathnilugal** [reply]
I have modified single particle model.
I will try to integrate into pybamm structure.
Are there any documents to add models into pybamm and use built in features.

**Reply by brosaplanella** [reply]
We don't have any documents explaining that, but I can guide you. What model have you implemented (or which changes have you made to SPM)?

**Reply by rtimms** [reply, marked as answer]
It's a little out of date (2 years ago) but we implemented some Li-S models as a separate library that worked with PyBaMM experiments. The trick is to define a `get_external_circuit_submodel` method in your model class so that current/voltage/power control can be handled correctly, see [here](https://github.com/DrMichaelCornish/LiS-Models/blob/master/models/base_lithium_sulfur_model.py#L25). I suspect the code would need updating to work with the current version of PyBaMM, but it should be a good starting point.
